Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,354 @@
+
+struct hostfxr_initialize_parameters
+{
+    size_t size;
+    const char_t * host_path;
+    const char_t * dotnet_root;
+};
+```
+
+The `hostfxr_initialize_parameters` structure stores parameters which are common to all forms of initialization.
+* `size` - the size of the structure. This is used for versioning. Should be set to `sizeof(hostfxr_initialize_parameters)`.
+* `host_path` - path to the native host (typically the `.exe`). This value is not used for anything by the hosting components. It's just passed to the CoreCLR as the path to the executable. It can point to a file which is not executable itself, if such file doesn't exist (for example in COM activation scenarios this points to the `comhost.dll`).
+* `dotnet_root` - path to the root of the .NET Core installation in use. This typically points to the install location from which the `hostfxr` has been loaded. For example on Windows this would typically point to `C:\Program Files\dotnet`. The path is used to search for shared frameworks and potentially SDKs.
+
+
+``` C
+int hostfxr_initialize_for_app(
+    int argc,
+    const char_t * argv[],
+    const char_t * app_path,
+    const hostfxr_initialize_parameters * parameters,
+    hostfxr_handle * host_context_handle
+);
+```
+
+Initializes the hosting components for running a managed application.
+When used to execute an app, the `app_path` (or CLI equivalent) will be used to locate the `.runtimeconfig.json` and the `.deps.json` which will be used to load the application and its dependent frameworks.
+* `argc` and `argv` - the command line - optional, if `argc` is `0` then `argv` is ignored.
+* `app_path` - path to the application (the managed `.dll`) to run. This can be `nullptr` if the app is specified in the command line arguments.
+* `parameters` - additional parameters - see `hostfxr_initialize_parameters` for details. (Could be made optional potentially)
+* `host_context_handle` - output parameter. On success receives an opaque value which identifies the initialized host context. The handle should be closed by calling `hostfxr_shutdown`.
+
+This function can only be called once per-process. It's not supported to run multiple apps in one process (even sequentially).
+
+This function will fail if there already is a CoreCLR running in the process.
+
+*Note: This is effectively a replacement for `hostfxr_main_startupinfo` and `hostfxr_main`. Currently it is not a goal to fully replace these APIs because they also support SDK commands which are special in lot of ways and don't fit well with the rest of the native hosting. There's no scenario right now which would require the ability to issue SDK commands from a native host. That said nothing in this proposal should block enabling even SDK commands through these APIs.*
+
+
+``` C
+int hostfxr_initialize_for_runtime_config(
+    const char_t * runtime_config_path,
+    const hostfxr_initialize_parameters * parameters,
+    hostfxr_handle * host_context_handle
+);
+```
```

Comment:
Is one intention of this API to provide a 'quick' check if an specific version is installed?  In that case, what output would indicate that the runtimeconfig.json would or would_not successfully find a runtime version?

Reply:
The main intent of this API is for the "load component" scenario. If the component has `.runtimeconfig.json` then it should point to that one - which would get the host the "component driven load" behavior. The host may provide a different one which would get it the "host driven load".

It can be used to check for framework resolution success as well (although it does more than that). The specific error codes are not defined in the document (too much detail), but the function will definitely fail if it can't resolve the frameworks requested by the runtime config. It would only return success error code if the framework resolution succeeded (among other things).